src/js/heuristicblocking.js

```diff
@@ -104,33 +113,143 @@ HeuristicBlocker.prototype = {
     let fqdn = (new URI(details.url)).host,
       origin = window.getBaseDomain(fqdn);
 
-    // if this is a main window request
+    // if this is a main window request, update tab data and quit
     if (details.type == "main_frame") {
-      // save the origin associated with the tab
-      log("Origin: " + origin + "\tURL: " + details.url);
       tabOrigins[details.tabId] = origin;
+      tabURLs[details.tabId] = details.url;
       return {};
     }
 
-    let tabOrigin = tabOrigins[details.tabId];
+    let tabOrigin = tabOrigins[details.tabId],
+      self = this;
 
     // ignore first-party requests
     if (!tabOrigin || origin == tabOrigin) {
       return {};
     }
 
     // abort if we already made a decision for this FQDN
-    let action = this.storage.getAction(fqdn);
+    let action = self.storage.getAction(fqdn);
     if (action != constants.NO_TRACKING && action != constants.ALLOW) {
       return {};
     }
 
-    // ignore if there are no tracking cookies
-    if (!hasCookieTracking(details, origin)) {
+    // check if there are tracking cookies
+    if (hasCookieTracking(details, origin)) {
+      self._recordPrevalence(fqdn, origin, tabOrigin);
       return {};
     }
 
-    this._recordPrevalence(fqdn, origin, tabOrigin);
+    if (checkForCookieShare && details.type == 'image') {
```

Comment:
Also check that "?" exists in the URL? Another chance to avoid retrieving cookies, etc.